Firestore/core/test/firebase/firestore/model/document_key_test.cc

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+
+#include <initializer_list>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+TEST(DocumentKey, Constructor_Empty) {
+  const DocumentKey default_key;
+  EXPECT_TRUE(default_key.path().empty());
+
+  const auto& empty_key = DocumentKey::Empty();
+  const auto& another_empty_key = DocumentKey::Empty();
+  EXPECT_EQ(default_key, empty_key);
+  EXPECT_EQ(empty_key, another_empty_key);
+  EXPECT_EQ(&empty_key, &another_empty_key);
+}
+
+TEST(DocumentKey, Constructor_FromPath) {
+  ResourcePath path{"rooms", "firestore", "messages", "1"};
+  const DocumentKey key_from_path_copy{path};
+  // path shouldn't have been moved from.
+  EXPECT_FALSE(path.empty());
+  EXPECT_EQ(key_from_path_copy.path(), path);
+
+  const DocumentKey key_from_moved_path{std::move(path)};
+  EXPECT_TRUE(path.empty());
+  EXPECT_FALSE(key_from_moved_path.path().empty());
+  EXPECT_EQ(key_from_path_copy.path(), key_from_moved_path.path());
+}
+
+TEST(DocumentKey, CopyAndMove) {
+  DocumentKey key({"rooms", "firestore", "messages", "1"});
+  const std::string path_string = "rooms/firestore/messages/1";
+  EXPECT_EQ(path_string, key.path().CanonicalString());
+
+  auto copied = key;
```

Comment:
style nit: It's best to avoid auto when its unclear what type is in use. But that statement leaves a lot up to the reader; after all, the type is just a few lines above. Personally, I typically only use auto when the type (more-or-less) otherwise appears on the same line, so I'd be inclined to not use auto here. But I don't much mind either way.

Optional.

(I'll refrain from pointing out subsequent instances of this. If you feel like adding the type here, then do so throughout.)

Reply:
Done.

I generally prefer `auto` when copying/moving if the "original" is in scope, because the type can only be the type of the "original", and also because it makes it clear that there is no conversion involved. But I don't feel strongly about it, so changed to explicit typing.

Reply:
The other case where auto is preferred is when the type is onerous or noisy (e.g. `std::vector<MyType>::const_iterator`).

This usage is just barely on the other side of what would occur to me to do but I don't care strongly.